### PR TITLE
auto-generate organization's msp crypto

### PR DESCRIPTION
**Pull request by 0xff-dev**

### CODE REVIEW

**Comment by 0xff-dev** [comment]
https://github.com/bestchains/fabric-operator/issues/14

File: pkg/offering/k8s/organization/organization.go [hunk 1]
```diff
@@ -81,3 +139,116 @@ func (organization *Organization) ReconcileManagers(instance *current.Organizati
 func (organization *Organization) CheckStates(instance *current.Organization) (common.Result, error) {
 	return organization.BaseOrganization.CheckStates(instance)
 }
+
+func (organization *Organization) enrollArg(ctx context.Context, namespace, name, enrollUser, enrollToken string) (*current.Enrollment, *CaProfile, error) {
+	caDeploy := v12.Deployment{}
+	var err error
+	// 字面量字符串都直接写了，先不管产量定义了
```
Comment (Abirdcfly) [inline pkg/offering/k8s/organization/organization.go:146]: may delete this

File: pkg/offering/k8s/organization/organization.go [hunk 2]
```diff
@@ -81,3 +139,116 @@ func (organization *Organization) ReconcileManagers(instance *current.Organizati
...
+			return
+		}
+
+		s.Data["signcert"] = resp.SignCert
+		s.Data["keystore"] = resp.Keystore
+		s.Data["org-ca-signcert"] = []byte(profileStr.CA.SignCerts)
+		s.Data["org-tlsca-signcert"] = []byte(profileStr.TLSCA.SignCerts)
+		s.Data["msg-config"] = []byte(mspConfig)
+
+		// ignore error
```
Comment (Abirdcfly) [inline pkg/offering/k8s/organization/organization.go:250]: IMO,use `_ = xxx()` to ignore error

File: controllers/organization/predict.go [hunk 3]
```diff
@@ -33,6 +33,7 @@ import (
 	"github.com/pkg/errors"
 	"gopkg.in/yaml.v2"
 	corev1 "k8s.io/api/core/v1"
+	errors2 "k8s.io/apimachinery/pkg/api/errors"
```
Comment (Abirdcfly) [inline controllers/organization/predict.go:36]: In other files, we use `k8serror`

Comment (bjwswang) [inline controllers/organization/predict.go:36]: should use `k8serrors` instead of `errors2`

File: api/v1beta1/organization_types.go [hunk 4]
```diff
@@ -43,6 +43,7 @@ type OrganizationSpec struct {
 	// Admin is the User/ServiceAccount with `Admin` role both in kubernetes and in CA
 	// +operator-sdk:gen-csv:customresourcedefinitions.specDescriptors=true
 	Admin string `json:"admin"`
+	Token string `json:"admintoken"`
```
Comment (bjwswang) [inline api/v1beta1/organization_types.go:46]: should use `AdminToken` to stay same with the json tag ?

File: controllers/organization/predict.go [hunk 5]
```diff
@@ -94,6 +95,9 @@ func (r *ReconcileOrganization) PredictOrganizationCreate(organization *current.
 	update.specUpdated = true
 	update.adminUpdated = true
 	update.clientsUpdated = true
+	if organization.Spec.Token != "" {
```
Comment (bjwswang) [inline controllers/organization/predict.go:98]: should do `tokenUpdated` check inner `if organization.HasType()` ?

File: controllers/organization/predict.go [hunk 6]
```diff
@@ -119,6 +123,23 @@ func (r *ReconcileOrganization) PredictOrganizationUpdate(oldOrg *current.Organi
...
+	// https://github.com/bestchains/fabric-operator/issues/14#issuecomment-1371948917
+	if oldOrg.Spec.Token != newOrg.Spec.Token {
+		update.tokenUpdated = update.adminUpdated
+		if !update.tokenUpdated {
+			secretExists := false
+			secret := corev1.Secret{}
+			err := r.client.Get(context.TODO(), types.NamespacedName{Name: oldOrg.GetName() + "-msg-crypto", Namespace: oldOrg.GetName()}, &secret)
+			if err != nil {
+				log.Error(err, fmt.Sprintf("get secret %s error", oldOrg.GetName()+"-msg-crypto"))
+				if errors2.IsNotFound(err) {
```
Comment (bjwswang) [inline controllers/organization/predict.go:135]: if `NotFound` ,means `secret not exists` ?

File: pkg/offering/k8s/organization/organization.go [hunk 7]
```diff
@@ -19,16 +19,71 @@
...
+
+type CATLS struct {
+	Cert string `json:"cert"`
+}
+
+type CACA struct {
+	SignCerts string `json:"signcerts"`
+}
+
+type CaProfile struct {
```
Comment (bjwswang) [inline pkg/offering/k8s/organization/organization.go:80]: CAConnectionProfile already defined at `https://github.com/bestchains/fabric-operator/blob/main/api/v1beta1/ibpca_types.go#L225`

File: pkg/offering/k8s/organization/organization.go [hunk 8]
```diff
@@ -81,3 +139,114 @@ func (organization *Organization) ReconcileManagers(instance *current.Organizati
...
+			CAHost: dstHost,
+			CAPort: dstPort,
+			CAName: name,
+			CATLS: &current.CATLS{
+				CACert: profile.TLS.Cert,
+			},
+			EnrollID:     enrollUser,
+			EnrollToken:  enrollToken,
+			EnrollUser:   enrollUser,
+			EnrollSecret: enrollUser,
```
Comment (bjwswang) [inline pkg/offering/k8s/organization/organization.go:195]: should add a comment to explain why `EnrollSecret` should use `enrollUser`

File: config/crd/bases/ibp.com_organizations.yaml [hunk 9]
```diff
@@ -365,6 +367,7 @@ spec:
                 type: object
             required:
             - admin
+            - admintoken
```
Comment (0xff-dev) [inline config/crd/bases/ibp.com_organizations.yaml:370]: `admintoken` should not be a required option.

**Review by bjwswang — APPROVED** [review]
LGTM

**Pull request merged**